pycue/opencue/cuelogging.py

```diff
@@ -0,0 +1,166 @@
+                    os.chmod(log_dir, 0o777)
+                # pylint: disable=broad-except
+                except Exception as e:
+                    # This is expected to fail when called in abq
+                    # But the directory should now be visible
+                    msg = e
+
+                if not os.access(log_dir, os.F_OK):
+                    err = "Unable to see log directory: %s, mkdir failed with: %s" % (
+                        log_dir, msg)
+                    raise RuntimeError(err)
+
+            if not os.access(log_dir, os.W_OK):
+                err = "Unable to write to log directory %s" % log_dir
+                raise RuntimeError(err)
+
+            try:
+                # Rotate any old logs to a max of MAX_LOG_FILES:
+                if os.path.isfile(self.filepath):
+                    rotateCount = 1
+                    while (os.path.isfile("%s.%s" % (self.filepath, rotateCount))
+                           and rotateCount < maxLogFiles):
+                        rotateCount += 1
+                    os.rename(self.filepath,
+                              "%s.%s" % (self.filepath, rotateCount))
+            # pylint: disable=broad-except
+            except Exception as e:
+                err = "Unable to rotate previous log file due to %s" % e
+                # Windows might fail while trying to rotate logs for checking if file is
+                # being used by another process. Frame execution doesn't need to
+                # be halted for this.
+                if platform.system() == "Windows":
+                    log.warning(err)
+                else:
+                    raise RuntimeError(err)
+            # pylint: disable=consider-using-with
+            self.fd = open(self.filepath, "w+", 1, encoding='utf-8')
+            try:
+                os.chmod(self.filepath, 0o666)
+            # pylint: disable=broad-except
+            except Exception as e:
+                err = "Failed to chmod log file! %s due to %s" % (self.filepath, e)
+                log.warning(err)
+        elif mode == MODE_READ:
+            self.fd = open(self.filepath, "r", encoding='utf-8')
+        else:
+            log.error("Unknown mode for log file!")
+
+    # pylint: disable=arguments-differ
+    def write(self, data, prependTimestamp=False):
+        """Abstract write function that will write to the correct backend"""
+        # Only work if in write mode, otherwise ignore
+        if self.mode == MODE_WRITE:
+            # Convert data to unicode
+            if isinstance(data, bytes):
+                data = data.decode('utf-8', errors='ignore')
+            if prependTimestamp is True:
+                lines = data.splitlines()
+                curr_line_timestamp = datetime.datetime.now().strftime("%H:%M:%S")
+                for line in lines:
+                    print("[%s] %s" % (curr_line_timestamp, line), file=self)
+            else:
+                self.fd.write(data)
```

Comment:
The function doesn't handle exceptions accordingly. A caller can be trying to write to a file with the wrong mode and nothing being written without an alert. I suggest raising an exception.

Reply:
Have split it into 2 separate classes, so should not be a concern anymore?